src/HDFFile.cxx

```diff
@@ -243,10 +318,104 @@ write_ds_numeric(hid_t hdf_parent, std::string name, std::vector<hsize_t> sizes,
   H5Pclose(dcpl);
 }
 
-static void write_ds_numeric_generic(std::string const &dtype, hid_t hdf_parent,
+static void
+write_ds_string(hid_t hdf_parent, std::string name, std::vector<hsize_t> sizes,
+                 std::vector<hsize_t> max, rapidjson::Value const *vals) {
+  size_t total_n = 1;
+  for (auto x : sizes) {
+    total_n *= x;
+  }
+  auto dcpl = H5Pcreate(H5P_DATASET_CREATE);
+  hid_t dsp = -1;
+  if (sizes.empty()) {
+    dsp = H5Screate(H5S_SCALAR);
+  } else {
+    dsp = H5Screate(H5S_SIMPLE);
+    H5Sset_extent_simple(dsp, (int)sizes.size(), sizes.data(), max.data());
+    if (max[0] == H5S_UNLIMITED) {
+      H5Pset_chunk(dcpl, sizes.size(), sizes.data());
+    }
+  }
+
+  std::vector<char const *> blob;
+  populate_string_pointers(blob, vals);
+
+  if (blob.size() != total_n) {
+    LOG(3, "error in sizes");
+    H5Sclose(dsp);
+    H5Pclose(dcpl);
+    return;
+  }
+
+  auto dt = H5Tcopy(H5T_C_S1);
+  H5Tset_size(dt, H5T_VARIABLE);
+  H5Tset_cset(dt, H5T_CSET_UTF8);
+
+  auto ds = H5Dcreate2(hdf_parent, name.data(), dt, dsp, H5P_DEFAULT, dcpl,
+                       H5P_DEFAULT);
+  auto err = H5Dwrite(ds, dt, H5S_ALL, H5S_ALL, H5P_DEFAULT, blob.data());
+  if (err < 0) {
+    LOG(3, "error while writing dataset");
```

Comment:
Are there different error codes such that you could give more specific error messages, or put the error code in the log message?